# Lista 2


### Zadanie 1
* Pełni wigoru powtórzymy zadanie 5 z listy 1. Następnie przeanalizujemy wpływ tak małej zmiany (jak zmiana ostatnich cyfr) na wynik.

In [1]:
### Identyczny kod jak ostatnio ###

function front_sum(A, B, size, breed) # frontward
    sum = breed(0.0)
    for i in 1:size
        sum += breed(A[i]) * breed(B[i])
    end
    return sum
end

function back_sum(A, B, size, breed) # backward
    sum = breed(0.0)
    for i in size:-1:1
        sum += breed(A[i]) * breed(B[i])
    end
    return sum
end

function sum_helper(NewA, breed) # help function, beacuse I am not sure how does Base.sum work
    sum = breed(0.0)
    for i in 1:size(NewA)[1]
        sum += NewA[i]
    end
    return sum
end

function smallest_to_biggest_sum(A, B, size, breed) 
    P = []
    N = []
    for i in 1:size
        mult = breed(A[i]) * breed(B[i]) # simulated dot product
        if mult >= 0 # positive mults go to positive array, same with negative, they go to negative
            append!(P, mult)
        else 
            append!(N, mult)
        end
    end
    P = sort(P, rev = true)
    N = sort(N, rev = true)
    positive = sum_helper(P, breed)
    negative = sum_helper(N, breed)
    return breed(positive + negative)
end

function biggest_to_smallest_sum(A, B, size, breed)
    P = []
    N = []
    for i in 1:size
        mult = breed(A[i]) * breed(B[i]) # simulated dot product
        if mult >= 0 # positive mults go to positive array, same with negative, they go to negative
            append!(P, mult)
        else 
            append!(N, mult)
        end
    end
    P = sort(P)
    N = sort(N)
    positive = sum_helper(P, breed)
    negative = sum_helper(N, breed)
    return breed(positive + negative)
end

biggest_to_smallest_sum (generic function with 1 method)

In [5]:
using LinearAlgebra
A = [2.718281828, -3.141592654, 1.414213562, 0.577215664, 0.301029995]
B = [1486.2497, 878366.9879, -22.37492, 4773714.647, 0.000185049]

println("by LinearAlgebra: ")
println("Float64: ", dot(A,B))
println("frontward sum:")
println("Float32: ",front_sum(A,B,5, Float32))
println("Float64: ",front_sum(A,B,5, Float64))
# println(front_sum(A,B,5, BigFloat))
println("backward sum:")
println("Float32: ",back_sum(A,B,5, Float32))
println("Float64: ",back_sum(A,B,5, Float64))
# println(back_sum(A,B,5, BigFloat))
println("biggest to smallest sum:")
println("Float32: ",biggest_to_smallest_sum(A,B,5, Float32))
println("Float64: ",biggest_to_smallest_sum(A,B,5, Float64))
println("smallest to biggest sum:")
println("Float32: ",smallest_to_biggest_sum(A,B,5, Float32))
println("Float64: ",smallest_to_biggest_sum(A,B,5, Float64))

by LinearAlgebra: 
Float64: -0.004296342739891585
frontward sum:
Float32: -0.4999443
Float64: -0.004296342739891585
backward sum:
Float32: -0.4543457
Float64: -0.004296342998713953
biggest to smallest sum:
Float32: -0.5
Float64: -0.004296342842280865
smallest to biggest sum:
Float32: -0.5
Float64: -0.004296342842280865


Wyniki z porpzedniej listy:

by LinearAlgebra: 

Float64: 1.0251881368296672e-10

frontward sum:

Float32: -0.4999443

Float64: 1.0251881368296672e-10

backward sum:

Float32: -0.4543457

Float64: -1.5643308870494366e-10

biggest to smallest sum:

Float32: -0.5

Float64: 0.0

smallest to biggest sum:

Float32: -0.5

Float64: 0.0

Wyniki zmieniły się tylko dla Float64 co jest zrozumiałe ze wzgledu na to, że prezcyzja liczb w wektorze jest większa niż przecyzja Float32. Na przykład:

In [7]:
Float32(0.5772156649)

0.5772157f0

Dla Float64 wyniki zmieniły się znacznie. Na przykład w przypadku standardowej sumy aż o 8 miejsc znaczących! Co dużo pokazuje o uwarunkowaniu zadania 5.

### Zadanie 2
* Pokażemy dziwne zachowanie wykresu funkcji $e^x\ln(1 + e^{-x})$ na trzech róznych oprogramowaniach do rysowania wykresów funkcji.
* Policzymy granicę funkcji.
* Wyjaśnimy jej nietypowe zachowanie.

wolfram alpha:
![wolfram](wolfram.png)

demos:
![demos](demos.png)

pyplot:
![pyplot](pyplot.png)

Jak widzimy wolfram i pyplot usilnie próbują uciąglić funkcję kiedy demos daje za wygraną.

Korzystając z reguły de l'Hospitala (i wolframa) wyliczamy, że $\lim_{x \to \infty} f(x) = 1$. Każdy wykres sugeruje, że granicą jest $0$. Bazując na wykresach widizmy, że dla $x\geq40$ $f(x) = 0$.

### Zadanie 4
*  Policzymy $|P(z_k)|$, $|p(z_k)|$ i $|z_k − k|$, gdzie $z_k$ to pierwiastek wielomianu $P$.

In [7]:
using Polynomials

natural = Polynomial(reverse([1, -210.0, 20615.0,-1256850.0,
      53327946.0,-1672280820.0, 40171771630.0, -756111184500.0,          
      11310276995381.0, -135585182899530.0,
      1307535010540395.0,     -10142299865511450.0,
      63030812099294896.0,     -311333643161390640.0,
      1206647803780373360.0,     -3599979517947607200.0,
      8037811822645051776.0,      -12870931245150988800.0,
      13803759753640704000.0,      -8752948036761600000.0,
      2432902008176640000.0]))
println(natural)
wilkinson = fromroots([i for i in 1:20])
println(wilkinson)

disturbed_wilkinson = Polynomial(reverse([1, -210.0 - 2^-23, 20615.0,-1256850.0,
      53327946.0,-1672280820.0, 40171771630.0, -756111184500.0,          
      11310276995381.0, -135585182899530.0,
      1307535010540395.0,     -10142299865511450.0,
      63030812099294896.0,     -311333643161390640.0,
      1206647803780373360.0,     -3599979517947607200.0,
      8037811822645051776.0,      -12870931245150988800.0,
      13803759753640704000.0,      -8752948036761600000.0,
      2432902008176640000.0]))
z_array = roots(natural) # getting natural's zeroes.


for (k , z) in enumerate(z_array)
    println("z_", k ," - ", k, ": ", abs(z - k))
    println("P (z_", k , ") = ", abs(natural(z)))
    println("P'(z_", k , ") = ", abs(disturbed_wilkinson(z)))
    println("p (z_", k , ") = ", abs(wilkinson(z)))
    println()
end

Polynomial(2.43290200817664e18 - 8.7529480367616e18*x + 1.3803759753640704e19*x^2 - 1.2870931245150988e19*x^3 + 8.037811822645051e18*x^4 - 3.599979517947607e18*x^5 + 1.2066478037803732e18*x^6 - 3.1133364316139066e17*x^7 + 6.30308120992949e16*x^8 - 1.014229986551145e16*x^9 + 1.307535010540395e15*x^10 - 1.3558518289953e14*x^11 + 1.1310276995381e13*x^12 - 7.561111845e11*x^13 + 4.017177163e10*x^14 - 1.67228082e9*x^15 + 5.3327946e7*x^16 - 1.25685e6*x^17 + 20615.0*x^18 - 210.0*x^19 + 1.0*x^20)
Polynomial(2432902008176640000 - 8752948036761600000*x - 4642984320068847616*x^2 + 5575812828558562816*x^3 + 8037811822645051776*x^4 - 3599979517947607200*x^5 + 1206647803780373360*x^6 - 311333643161390640*x^7 + 63030812099294896*x^8 - 10142299865511450*x^9 + 1307535010540395*x^10 - 135585182899530*x^11 + 11310276995381*x^12 - 756111184500*x^13 + 40171771630*x^14 - 1672280820*x^15 + 53327946*x^16 - 1256850*x^17 + 20615*x^18 - 210*x^19 + x^20)
z_1 - 1: 3.0109248427834245e-13
P (z_1) = 35696.50964788257


Jak zwykle rozbieżność w wynikach pochodzi z dokładności z jaką możemy reprezentować liczby rzeczywiste w dość rzadkim zbiorze liczb maszynowych. Każdy błąd, nawet najmniejszy, powiela się na kolejne operacje. Widzimy, że różnica pomiędzy $z_1$ i $1$ jest minimalna, niemniej wartość wyrażenia $P(z_1) = 35696.50964788257$, powinno być $0$. Innymi słowo, 13 cyfra po przecinku wygerenowała wynik bardzo odległy od spodziewanego w wielomianie którego współczynniki są typu zmiennoprzecinkowego. Dużo gorzej radzi sobie za to wielomian $p$, którego wpółczynniki są trzymane w zmiennych typu całkowitego. Dla większości wartości argumentu otrzymujemy liczby o 20 cyfrach powyżej zera.

Wielomian $P'$ mający odrobinę zmieniony współczynnik przy $x^{19}$ zaczyna się różnić od $P$ już przy $z_5$. Przekazując coraz większe liczby jako "miejsca zerowe" skala błędów zaczyna się coraz bardziej nawarstwiać. Dla $z_{20}$ już dla wszystkich wielomianów dostajemy liczby rzędu $10^{13}$ i większe.